From 14441ebd8f6156d6c3f78149168e76f79246fd1c Mon Sep 17 00:00:00 2001
From: Ulf Hermann <ulf.hermann@qt.io>
Date: Wed, 3 Sep 2025 14:50:28 +0200
Subject: [PATCH 007/239] QtQuick: Drop redundant members from QQuickRepeater

We only need to store the model once. The object stored that way is
guaranteed to either be the model that was assigned or hold its
contents.

Task-number: QTBUG-139941
Change-Id: Icec9fd9f756042c44a3e3189bf5c7ee4253b086f
Reviewed-by: Sami Shalayel <sami.shalayel@qt.io>
(cherry picked from commit bc28a90abbb017d72930e0760c9852518d3727b1)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/quick/items/qquickrepeater.cpp   | 31 ++++++++++++++--------------
 src/quick/items/qquickrepeater_p_p.h |  3 ---
 2 files changed, 15 insertions(+), 19 deletions(-)

diff --git a/src/quick/items/qquickrepeater.cpp b/src/quick/items/qquickrepeater.cpp
index e5a3a77a34..c84de58c5b 100644
--- a/src/quick/items/qquickrepeater.cpp
+++ b/src/quick/items/qquickrepeater.cpp
@@ -16,7 +16,6 @@ QT_BEGIN_NAMESPACE
 QQuickRepeaterPrivate::QQuickRepeaterPrivate()
     : model(nullptr)
     , ownModel(false)
-    , dataSourceIsObject(false)
     , delegateValidated(false)
     , explicitDelegate(false)
     , explicitDelegateModelAccess(false)
@@ -152,12 +151,11 @@ QVariant QQuickRepeater::model() const
 {
     Q_D(const QQuickRepeater);
 
-    if (d->dataSourceIsObject) {
-        QObject *o = d->dataSourceAsObject;
-        return QVariant::fromValue(o);
-    }
-
-    return d->dataSource;
+    if (d->ownModel)
+        return static_cast<QQmlDelegateModel *>(d->model.data())->model();
+    if (d->model)
+        return QVariant::fromValue(d->model.data());
+    return QVariant();
 }
 
 void QQuickRepeater::setModel(const QVariant &m)
@@ -167,20 +165,20 @@ void QQuickRepeater::setModel(const QVariant &m)
     if (model.userType() == qMetaTypeId<QJSValue>())
         model = model.value<QJSValue>().toVariant();
 
-    if (d->dataSource == model)
+    QQmlDelegateModelPointer oldModel(d->model);
+    if (d->ownModel) {
+        if (oldModel.delegateModel()->model() == model)
+            return;
+    } else if (QVariant::fromValue(d->model) == model) {
         return;
+    }
 
     clear();
 
-    QQmlDelegateModelPointer oldModel(d->model);
     d->disconnectModel(this, &oldModel);
-
     d->model = nullptr;
-    d->dataSource = model;
 
     QObject *object = qvariant_cast<QObject *>(model);
-    d->dataSourceAsObject = object;
-    d->dataSourceIsObject = object != nullptr;
 
     QQmlDelegateModelPointer newModel(qobject_cast<QQmlInstanceModel *>(object));
     if (newModel) {
@@ -217,10 +215,11 @@ void QQuickRepeater::setModel(const QVariant &m)
         }
         d->model = newModel.instanceModel();
     } else if (d->ownModel) {
+        // d->ownModel can only be set if the old model is a QQmlDelegateModel.
+        Q_ASSERT(oldModel.delegateModel());
         newModel = oldModel;
         d->model = newModel.instanceModel();
-        if (QQmlDelegateModel *delegateModel = newModel.delegateModel())
-            delegateModel->setModel(model);
+        newModel.delegateModel()->setModel(model);
     } else {
         newModel = QQmlDelegateModel::createForView(this, d);
         if (d->explicitDelegate) {
@@ -534,7 +533,7 @@ void QQuickRepeater::initItem(int index, QObject *object)
         // If the item comes from an ObjectModel, it might be used as
         // ComboBox/Menu/TabBar's contentItem. These types unconditionally cull items
         // that are inserted, so account for that here.
-        if (d->dataSourceIsObject)
+        if (d->model && !d->ownModel)
             QQuickItemPrivate::get(item)->setCulled(false);
         if (index > 0 && d->deletables.at(index-1)) {
             item->stackAfter(d->deletables.at(index-1));
diff --git a/src/quick/items/qquickrepeater_p_p.h b/src/quick/items/qquickrepeater_p_p.h
index 2d24f47be4..2ce7e4453c 100644
--- a/src/quick/items/qquickrepeater_p_p.h
+++ b/src/quick/items/qquickrepeater_p_p.h
@@ -53,10 +53,7 @@ private:
     void disconnectModel(QQuickRepeater *q, QQmlDelegateModelPointer *model);
 
     QPointer<QQmlInstanceModel> model;
-    QVariant dataSource;
-    QPointer<QObject> dataSourceAsObject;
     bool ownModel : 1;
-    bool dataSourceIsObject : 1;
     bool delegateValidated : 1;
     bool explicitDelegate : 1;
     bool explicitDelegateModelAccess : 1;
-- 
2.51.2

